common/lib/xmodule/xmodule/vertical_block.py

```diff
@@ -107,14 +64,24 @@ def student_view(self, context):
             user_service = self.runtime.service(self, 'user')
             child_context['username'] = user_service.get_current_user().opt_attrs['edx-platform.username']
 
+        child_blocks = self.get_display_items()
+
+        child_blocks_to_complete_on_view = set()
         completion_service = self.runtime.service(self, 'completion')
+        if completion_service and completion_service.completion_tracking_enabled():
+            child_blocks_to_complete_on_view = blocks_to_mark_complete_on_view(completion_service, child_blocks)
 
         child_context['child_of_vertical'] = True
         is_child_of_vertical = context.get('child_of_vertical', False)
 
         # pylint: disable=no-member
-        for child in self.get_display_items():
-            rendered_child = child.render(STUDENT_VIEW, child_context)
+        for child in child_blocks:
+            child_block_context = copy(child_context)
+            if child in child_blocks_to_complete_on_view:
+                child_block_context['wrap_xblock_data'] = {
+                    'mark-completed-on-view-after-delay': completion_service.get_completion_by_viewing_delay_ms()
```

Comment:
nit: `completion_service.get_completion_by_viewing_delay_ms()` will return the same thing on each call, so you could just store it in a variable outside your `for` loop.

Reply:
I have updated this.